### PR TITLE
feat: support casting in globals

### PR DESCRIPTION
**Pull request by michaeljklein**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/5160

## Summary\*

Adds a case for casts in comptime globals

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir/comptime/value.rs [hunk 1]
```diff
@@ -174,6 +174,47 @@ impl Value {
...
+    /// Returns `None` for negative integers.
+    pub(crate) fn to_u128(&self) -> Option<u128> {
+        match self {
+            Self::Field(value) => Some(value.to_u128()),
+            Self::I8(value) => {
+                if value < &0 {
+                    None
+                } else {
+                    Some(*value as u128)
+                }
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/comptime/value.rs:188]:
```suggestion
                (*value >= 0).then(|| *value as u128)
```
Could save some lines here

File: compiler/noirc_frontend/src/hir/resolution/errors.rs [hunk 2]
```diff
@@ -94,6 +94,8 @@ pub enum ResolverError {
     NoPredicatesAttributeOnUnconstrained { ident: Ident },
     #[error("#[fold] attribute is only allowed on constrained functions")]
     FoldAttributeOnUnconstrained { ident: Ident },
+    #[error("Interpreting array-length failed")]
+    ArrayLengthInterpreter { error: InterpreterError, span: Span },
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/resolution/errors.rs:98]: Can you rename this error to something which doesn't reveal/specify to users that we're interpreting array lengths? E.g. `InvalidArrayLength` or similar?

File: compiler/noirc_frontend/src/hir/resolution/errors.rs [hunk 3]
```diff
@@ -386,6 +388,11 @@ impl<'a> From<&'a ResolverError> for Diagnostic {
                 diag.add_note("The `#[fold]` attribute specifies whether a constrained function should be treated as a separate circuit rather than inlined into the program entry point".to_owned());
                 diag
             }
+            ResolverError::ArrayLengthInterpreter { error, span } => Diagnostic::simple_error(
+                "Interpreting array length failed".into(),
+                format!("with error: {}", Diagnostic::from(error)),
+                *span,
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/resolution/errors.rs:394]: Same here - maybe we should just display the error directly?

**Review by jfecher — CHANGES_REQUESTED** [review]
Looks like we're missing the elaborator code for this

File: compiler/noirc_frontend/src/hir/resolution/errors.rs (hunk 3 above)
Comment (TomAFrench) [reply compiler/noirc_frontend/src/hir/resolution/errors.rs:394]: Agreed on this. Ideally we'd surface the original error directly or at least just delegate to it.

File: compiler/noirc_frontend/src/hir/resolution/resolver.rs [hunk 4]
```diff
@@ -2064,6 +2068,21 @@ impl<'a> Resolver<'a> {
...
+                    Interpreter::evaluate_cast_one_step(&cast, rhs, lhs_value, self.interner)
+                        .map_err(|error| {
+                            Some(ResolverError::ArrayLengthInterpreter { error, span })
+                        })?;
+
+                if let Some(result) = evaluated_value.to_u128() {
+                    Ok(result)
+                } else {
+                    Err(Some(ResolverError::InvalidArrayLengthExpr { span }))
+                }
```
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/resolution/resolver.rs:2084]:
```suggestion
                evaluated_value.to_u128().ok_or_else(|| Some(ResolverError::InvalidArrayLengthExpr { span }))
```

File: compiler/noirc_frontend/src/hir/comptime/errors.rs [hunk 5]
```diff
@@ -5,7 +5,7 @@ use noirc_errors::{CustomDiagnostic, Location};
 use super::value::Value;
 
 /// The possible errors that can halt the interpreter.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Eq)]
```
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/comptime/errors.rs:8]: We don't actually use this trait on `TypeCheckError` (which causes the requirement for it here). We could just remove it from the other errors for now.

**Review by jfecher — APPROVED** [review]
LGTM

**Pull request merged**